DDP-5402 - implemented an API ActivityInstanceFromDefinitionBuilder for ActivityInstance creation

evgeniipr opened the pull request:
## Context

It is implemented an API for `ActivityInstance` creation from a study definition stored in `ActivityDefStore`. 
The instance data (like activity instance, answers) are fetched from DB. Validation messages also fetched from DB but once it's fetched it is cached in `ActivityDefStore` cache and next time is taken from the cache.

The main class of this API is `ActivityInstanceFromDefinitionBuilder`. It has one method:
```
public Optional<ActivityInstance> buildActivityInstance(
            Handle handle,
            String userGuid,
            String operatorGuid,
            String studyGuid,
            String instanceGuid,
            ContentStyle style,
            String isoLangCode
    );
```

This API currently used in `GetActivityInstanceRoute` instead of old `ActivityInstanceService.getTranslatedActivity()`:
```
 Optional<ActivityInstance> inst = actInstService.buildInstanceFromDefinition(
                handle, userGuid, operatorGuid, studyGuid, instanceGuid, style, isoLangCode
 );
```
The performance tests indicated about 3.5 times of speed increase of `ActivityInstance` reading in endpoint  `GetActivityInstanceRoute`.

The next step will be: remove old implementation `ActivityInstanceService.getTranslatedActivity()` from the code (subtask https://broadinstitute.atlassian.net/browse/DDP-5758).


## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [x] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [x] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [x] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

Commit:
DDP-5402 - implemented a service (ActivityInstanceAssembleService) used to build ActivityInstance object from ActivityDefStore data which already contains most of activity instance data.

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstanceassembler/FormInstanceCreator.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    private void render(FormInstance formInstance, FormActivityDef formActivityDef) {
+        // TODO
```
@yufengwng could you advise about rendering of `FormInstance` templates?
There is a method `FormInstance.renderContent()` where rendered `readonlyHint`, `activityDefinitionLastUpdatedText`.  What is the best way to render this strings then without excessive calls to DB?
I believe it is not possible to avoid DB calls in this case. What do you think?
Possibly it needs to update `ActivityDefStore`  
And another question about rendering: some of `..Def` classes contains only templateID-s (and `Template` objects are not assigned - i.e. nulls): does it mean that it needs to query DB to get `Template` data? I didn't find any cache (map) with templates from where it could be fetched by IDs.

yufengwng replied:
I'm not sure I have all the answers, but here's my recommendation:

* Try updating the various DAOs (such as `ContentBlockDao::colleckBlockDefs()`) so that they query out all the `Template` objects. This way, we'll have what we need from the `ActivityDef` objects. It does add cost but this is only done once and then it'll be cached in `ActivityDefStore`.

* Try adding a new overloaded `render()` method to `Template` class that takes in a ` Map<String, Object> context`. This way, we can pass in a `RenderValueProvider` object, just like how things are done in `FormInstance::renderContent()`.

* Lastly, instead of calling `FormInstance::renderContent()` or rendering templates with `bulkRender()` and template ids, take advantage of the fact that we have `Template` objects that we can render from. For things like `readonlyHint` and `lastUpdatedText`, maybe try to add new constructor that takes in a string instead of the template id.

Commit:
DDP-5402 - refactored ActivityInstance assemble service; provided templates more better rendering.

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/ContentBlockDao.java:
```diff
@@ -89,11 +90,22 @@ default void disableContentBlock(long blockId, RevisionMetadata meta) {
+            }
+        });
+        Map<Long, Template> templates = getTemplateDao().collectTemplatesByIdsAndTimestamp(templateIds, timestamp);
+
```
@yufengwng please check this change: here is added a ContentBlock templates collecting, and then the collected templates are passed to a ContentBlock constructor (see below).

yufengwng replied:
Looks about right.

evgeniipr commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstanceassembler/FormInstanceCreator.java:
```diff
@@ -0,0 +1,106 @@
+                null);
+    }
+
+    private void render(FormInstance formInstance) {
```
@yufengwng in this method rendered `readonlyHint` and `activityDefinitionLastUpdatedText`. Currently the result is the same as during ActivityInstance normal reading (the texts are identical). But yet I'm not sure if it's correct from the versions perspective etc.. I need to check still. And as you see here I just call template.render(0 and then HtmlConverter. So it looks quite similar to formInstance.renderContent() logic but I think I need more deeper understanding of `activityDefinitionLastUpdated` rendering (because it depends on `activityDefinitionLastUpdated'). So, this logic I definitely need to fix in my code still.

yufengwng replied:
Take a look at `FormActivityDao::findDefByDtoAndVersion()`. When we build a `FormActivityDef` object, we are already querying for the appropriate version. So the read-only hint and last updated text you get from the `FormActivityDef` object is what you should use.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstanceassembler/ActivityInstanceAssembleService.java:
```diff
@@ -0,0 +1,188 @@
+ * <p>NOTE: it is defined a class {@link Context} used to pass the basic parameters to each element creator (so it no
+ * need to pass multiple parameter in each creator - only one parameter.
+ */
+public class ActivityInstanceAssembleService {
```
Can we just put this in `ActivityInstanceService` instead of creating a new class?

evgeniipr replied:
I renamed `ActivityInstanceAssembleService` to `ActivityInstanceFromActivityDefStoreBuilder` and added to `ActivityInstanceService` a method:
```
public Optional<ActivityInstance> buildInstanceFromActivityDefStore(
            Handle handle,
            String isoLangCode,
            ContentStyle style,
            String studyGuid,
            String userGuid,
            String operatorGuid,
            ActivityInstanceDto instanceDto) {
        return new ActivityInstanceFromActivityDefStoreBuilder().buildActivityInstance(
                handle, isoLangCode, style, studyGuid, userGuid, operatorGuid, instanceDto
        );
}
```

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstanceassembler/FormInstanceCreator.java:
```diff
@@ -0,0 +1,106 @@
+                instanceDto.getActivityCode(),
+                formActivityDef.getFormType(),
+                instanceDto.getGuid(),
+                activityI18nDetail != null ? activityI18nDetail.getTitle() : null,
```
The `FormActivityDef` object should already have things like title and subtitle. I think we should reuse those instead of querying for an `ActivityI18nDetail`. Take a look at `FormActivityDao::findDefByDtoAndVersion()` method -- it already queries for those `ActivityI18nDetail`. I think what you need to do here is iterate through those lists and pick out the right one using the language code.

Commit:
DDP-5402 - refactored: renamed module package to 'activityinstancebuilder' and main class to 'ActivityInstanceFromActivityDefStoreBuilder'; fixed templates detection from 'ActivityDefStore'; add a method which starts ActivityInstance building to 'ActivityInstanceService'.

Commit:
DDP-5402 - fix errors in ActivityInstance generation

Commit:
Merge branch 'develop' into DDP-5402-instance-data-read-optimization

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/definition/ComponentBlockDef.java:
```diff
@@ -26,6 +26,9 @@
     @SerializedName("hideNumber")
     private boolean hideNumber;
 
+    @SerializedName("revisionId")
+    private long revisionId;
```
Note that these definition objects are what we use to deserialize study-builder activity configurations in order to insert them into the database (check out `ActivityBuilder` in study-builder codebase). `revisionId` is not something we'll deserialize from the activity configuration JSON. How about we change the following so it doesn't play a part in JSON?

```
private transient long revisionId;
```

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/ActivityInstanceService.java:
```diff
@@ -1,19 +1,5 @@
-import java.util.function.Function;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
-
```
Order of imports here is not consistent with Java files in this repo. Can you fix your code editor settings so that the order of imports is:
```
import static

import java.*
import javax.*

<third part imports>
```

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/ActivityInstanceService.java:
```diff
@@ -563,4 +531,21 @@ public void deleteInstance(Handle handle, ActivityInstanceDto instanceDto) {
+     * Build {@link ActivityInstance} from data cached stored in {@link ActivityDefStore}.
+     * Some of data (answers, rule messages) are queried from DB.
+     */
+    public Optional<ActivityInstance> buildInstanceFromActivityDefStore(
```
Consider renaming the method. The focus here is that we're rendering the activity instance from its definition. It just so happens that the definition is stored in a class called ActivityDefStore. The definition could very well be stored somewhere else (like Redis).

```
buildInstanceFromDefinition()

ActivityInstanceFromDefinitionBuilder()
```

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/FormInstanceCreator.java:
```diff
@@ -0,0 +1,108 @@
+        String subtitle = extractOptionalActivityTranslation(formActivityDef.getTranslatedSubtitles(), context.getIsoLangCode());
+
+        boolean readonly = ActivityInstanceUtil.isReadonly(
+                context.getActivityInstanceDto().getEditTimeoutSec(),
```
`editTimeoutSed` is a property of definition. Use `ActivityDef::getEditTimeoutSec()`.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/FormInstanceCreator.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    private FormInstance constructFormInstance() {
+        ActivityInstanceDto instanceDto = context.getActivityInstanceDto();
```
Have you considered passing in the whole `FormResponse` object instead of both a `ActivityInstanceDto` and list of answers? FormResponse is missing `isHidden` and `sectionIndex`, but once we add those, it should give you everything you need from ActivityInstanceDto.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/FormInstanceCreator.java:
```diff
@@ -0,0 +1,108 @@
+    public FormInstance createFormInstance() {
+        FormInstance formInstance = constructFormInstance();
+        addChildren(formInstance);
+        updateBlockStatuses(formInstance);
```
Tracing through the current code,  looks like`updateBlockStatuses()` is the last thing we call.
```
1. new instance
2. load sections
3. render
4. set display numbers
5. update block statuses
```

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/FormInstanceCreator.java:
```diff
@@ -0,0 +1,108 @@
+                null);
+    }
+
+    private void render(FormInstance formInstance) {
```
As you're aware, this is not feature-parity with existing render method. We need to create a `RenderValueProvider` and pass that into the `Template` so it can use it as rendering context. For last updated text, we also need to pass `activityDefinitionLastUpdated` into the rendering context map. Please see the original method for more details.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/FormSectionCreator.java:
```diff
@@ -0,0 +1,38 @@
+        if (formSectionDef != null) {
+            FormSection formSection = constructFormSection(formSectionDef);
+            addChildren(formSection, formSectionDef);
+            applyRenderedTemplates(formSection);
```
Can we use the name template object rather than calling `applyRenderedTemplates()`?

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/FormInstanceCreator.java:
```diff
@@ -0,0 +1,108 @@
+                instanceDto.getStatusType() != null ? instanceDto.getStatusType().name() : null,
+                readonly,
+                formActivityDef.getListStyleHint(),
+                getTemplateId(formActivityDef.getReadonlyHintTemplate()),
```
Hmmm, I'm not sure about calling `getTemplateId()` -- it's doing extra work. The implementation of `getTemplateId()` calls `template.render()`, which means we're rendering the read-only hint twice.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/SectionIconCreator.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    private SectionIcon constructSectionIcon(SectionIcon sectionIconDef) {
+        SectionIcon sectionIcon = new SectionIcon(
```
Hmmm, they're the same thing..?

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/block/impl/ContentBlockCreator.java:
```diff
@@ -0,0 +1,24 @@
+    public ContentBlock createContentBlock(ContentBlockDef contentBlockDef) {
+        return new ContentBlock(
+                getTemplateId(contentBlockDef.getTitleTemplate()),
+                getTemplateId(contentBlockDef.getBodyTemplate())
```
Note that the whole approach with template ids and `registerTemplateIds`/`applyRenderedTemplates` was a workaround to optimize rendering templates when we didn't have the definition objects. But now that we're rendering directly from activity definition objects, I recommend we take full advantage of those template objects.

For example, right here we can create a new `ContentBlock` constructor that takes in the rendered title/body strings rather than ids. If you take a look at the code, no one is calling `getTitleTemplateId`/`getBodyTemplateId`, so we can probably just get rid of those entirely.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/block/impl/GroupBlockCreator.java:
```diff
@@ -0,0 +1,40 @@
+    private void addChildren(GroupBlock groupBlock, GroupBlockDef groupBlockDef) {
+        if (groupBlockDef.getNested() != null) {
+            groupBlockDef.getNested().forEach(b ->
+                    groupBlock.getNested().add(new FormBlockCreator(context).createBlock(b)));
```
Move the `new FormBlockCreator()` outside the loop so we're not new-ing up a bunch of objects.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/block/question/ValidationRuleCreator.java:
```diff
@@ -0,0 +1,170 @@
+
+    private String resolveRuleMessage(RuleDef ruleDef) {
+        ValidationDao validationDao = context.getHandle().attach(ValidationDao.class);
+        return validationDao.getJdbiI18nValidationMsgTrans().getValidationMessage(
```
I think we can cache these default validation message texts. This is probably a good time to do it.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/block/question/QuestionUtil.java:
```diff
@@ -0,0 +1,14 @@
+
+public class QuestionUtil {
+
+    public static boolean isReadOnly(ActivityInstanceFromActivityDefStoreBuilder.Context context, QuestionDef questionDef) {
```
This is not feature-parity with existing calculation of question read-only. Please see `ddp.util.QuestionUtil` class. Also, I recommend removing this class here and add a new method in the other util class -- having two classes with same name is a bit confusing.
```
isReadOnly(Handle handle, QuestionDef questionDef, InstanceStatusType instanceStatus, long instanceId)
```

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/block/question/impl/BoolQuestionCreator.java:
```diff
@@ -0,0 +1,37 @@
+        super(context);
+    }
+
+    public BoolQuestion createBoolQuestion(QuestionCreator questionCreator, BoolQuestionDef questionDef) {
```
For simple question types like these, perhaps we can move these construction methods into `QuestionCreator` itself? I'll rather have a bigger class file than a ton of tiny files.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/block/question/impl/CompositeQuestionCreator.java:
```diff
@@ -0,0 +1,54 @@
+
+    private List<Question> getChildQuestions(CompositeQuestionDef questionDef) {
+        List<Question> childQuestions = new ArrayList<>();
+        QuestionCreator questionCreator = new QuestionCreator(context);
```
Is it important to create a new object? We already have a `QuestionCreator` we can use.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/ElementCreator.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Abstract creator class
+ */
+public class ElementCreator {
```
Is there advantage to having this `ElementCreator`? I don't see how having this parent class helps us. If we really need these, I think `getTemplateId()` and `applyRenderedTemplates()` could be implemented as methods on the `Context` class itself.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/route/GetActivityInstanceRoute.java:
```diff
@@ -127,6 +125,27 @@ public Object handle(Request request, Response response) {
+        Optional<ActivityInstance> inst = actInstService.buildInstanceFromActivityDefStore(
+                handle, isoLangCode, style, studyGuid, userGuid, operatorGuid, instanceDto
+        );
+        if (inst.isEmpty()) {
+            inst = actInstService.getTranslatedActivity(
+                    handle, userGuid, operatorGuid, instanceDto.getActivityType(), instanceGuid, isoLangCode, style
+            );
+        }
```
I don't think we should fallback to the old approach here. Once we switch to new approach, we should deprecate this. I look forward to deleting code for `getTranslatedActivity()` and related methods!

Commit:
DDP-5402 - fix code in response to PR comments: rename some methods; redesign templates rendering

Commit:
DDP-5402 - refactoring: join smaller classes (creating AI components) into bigger once; fix Question.isReadOnly logic

evgeniipr commented:
@yufengwng I pushed next commit: 
- removed small question/block creation classes and created 2 bigger classes instead;
- did some optimization (all Creators' construction - at the beginning of the process); 
- fixed Question.isReadOnly detection.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/FormResponse.java:
```diff
@@ -36,9 +36,12 @@ public FormResponse(@ColumnName("instance_id") long id,
                         @ColumnName("activity_code") String activityCode,
                         @ColumnName("activity_version_tag") String activityVersionTag,
+                        @ColumnName("is_hidden") Boolean isHidden,
+                        @ColumnName("sectionIndex") Integer sectionIndex,
```
JDBI is smart enough to work with both snake_case and camelCase, but let's be consistent here.
```suggestion
                        @ColumnName("section_index") Integer sectionIndex,
```

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/AbstractCreator.java:
```diff
@@ -0,0 +1,26 @@
+        this.context = context;
+    }
+
+    protected Long renderTemplateIfDefined(Template template) {
```
I still think this can be a util method on the `context` object itself, rather than needing to share an abstract class. Or maybe put it in `TemplateRenderUtil` if you want to keep methods out of `context`.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/block/question/QuestionCreator.java:
```diff
@@ -0,0 +1,80 @@
+    <T extends Answer> List<T> getAnswers(Class<T> type, String questionStableId) {
+        List<T> answers = new ArrayList<>();
+        if (context.getFormResponse().getAnswers() != null) {
+            answers = context.getFormResponse().getAnswers().stream()
```
You can lookup the answer by questionStableId instead of streaming it. The choice of using a `List` of answers was a legacy design decision -- there should only be one answer per question.

Commit:
DDP-5402 - refactor ActivityInstance builder (remove base AbstractCreator); fix a problem which occurs when ActivityDefStore still empty (load it with data).

Commit:
DDP-5402 - fix incorrect name in SQL mapping

evgeniipr commented:
> I still think this can be a util method on the context object itself, rather than needing to share an abstract class.

I removed the abstract class.

evgeniipr commented:
> You can lookup the answer by questionStableId instead of streaming it. 

Fixed. I also marked the `FormResponse.getAnswers()` as `deprecated`.

evgeniipr commented:
> JDBI is smart enough to work with both snake_case and camelCase, but let's be consistent here.

Fixed

Commit:
DDP-5402 - populate definition cache if activity data not found in ActivityDefStore;
read only one answer for a question (and mark a method returning a list of answers as deprecated);
store validation rules messages in ActivityDefStore cache

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/db/ActivityDefStore.java:
```diff
@@ -161,6 +166,15 @@ public void setActivityDef(String studyGuid, String activityCode, String version
         }
     }
 
+    public String findValidationRuleMessageMap(Handle handle, RuleType ruleType, long langCodeId) {
```
Can you rename this method? We're not finding a "Map" but rather just the rule message.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/FormResponse.java:
```diff
@@ -36,15 +36,28 @@ public FormResponse(@ColumnName("instance_id") long id,
+     *       The choice of using a List of answers was a legacy design decision -
+     *       there should only be one answer per question.
+     */
+    @Deprecated
```
I think there's a misunderstanding here. This method should not be deprecated. When I said "legacy design decision", I was not talking about this method -- what I meant was that if you looked at the `Question` class it has a `List<Answer>`. Question should have one answer, so this `List<Answer>` was an old design decision (and no, we do not need to refactor that right now!).

And note that this `getAnswers()` method does NOT return a list of answers for a single question, rather it returns all the answers for all questions in the activity instance. Which is why I suggested not to use this one -- since we want the answer for a single question, we should use `getAnswers(String questionStableId)` instead.

Sorry if I've caused any confusions! Can you please remove the `@Deprecated` annotation?

Commit:
DDP-5402 - rename methods; add comments; fix logic of AI builder

Commit:
DDP-5402 - provide section icons reading during ActivityDefStore populating with data;
- fix FormInstance title/subTitle rendering;
- fix validation rule messages detection;
- fix field 'is_hidden' mapping (add to an SQL result list);
- add timing measurement to check the speed of the new ActivityInstance reading code;
- fix data copying for PickList group.

Commit:
DDP-5402 - add comments; refactoring: rename some classes.

Commit:
Configuration to deploy FB to dev

Commit:
Merge branch 'develop' into DDP-5402-instance-data-read-optimization

Commit:
DDP-5402 - optimize templates render context creation

Commit:
DDP-5402 - change timing messages to debug level of logging; add comments marking old AI building method as deprecated (as a new one is implemented)

Commit:
DDP-5402 - fix formatting error

Commit:
DDP-5402 - re-implement unit tests for a new version of AI builder (ActivityInstanceService.getTranslatedActivity() usage removed from tests); fix templates rendering problems

Commit:
DDP-5402 - remove unused variables; add comments

Commit:
DDP-5402 - fix blocks creation: if block is marked as "deprecated" then it should not be visible

Commit:
DDP-5402 - added unit test for testing deprecated questions; fixed error in passing ActivityInstance parameters

Commit:
Merge branch 'develop' into DDP-5402-instance-data-read-optimization

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/ActivityInstanceFromDefinitionBuilder.java:
```diff
@@ -0,0 +1,121 @@
+            ContentStyle style,
+            String isoLangCode
+    ) {
+        LOG.info("Start ActivityInstance building from definition (ActivityDefStore). StudyGuid={}, instanceGuid={}",
```
Might want to change these to `debug()` logs -- there's going to be alot of activity instance fetching, so we don't want to clog up production logs! Likewise with the `SUCCESSFULLY` log below.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/ActivityInstanceFromDefinitionBuilder.java:
```diff
@@ -0,0 +1,121 @@
+                studyGuid, instanceGuid);
+        var formResponse = ActivityInstanceUtil.getFormResponse(handle, instanceGuid);
+        if (formResponse.isEmpty()) {
+            LOG.warn("Error reading form activity by guid=" + instanceGuid);
```
There's already logging for the `isEmpty()` case in `GetActivityInstanceRoute` so this warning here is a bit redundant. In general, probably good idea to let caller deal with logging/reporting.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/db/ActivityDefStore.java:
```diff
@@ -161,6 +182,37 @@ public void setActivityDef(String studyGuid, String activityCode, String version
+
+    private String detectValidationRuleMessage(
+            Handle handle, RuleType ruleType, Long hintTemplateId, long langCodeId, long timestamp) {
+        String correctionHint = null;
+        if (hintTemplateId != null) {
+            correctionHint = i18nContentRenderer.renderContent(handle, hintTemplateId, langCodeId, timestamp);
+        }
+        if (correctionHint != null) {
+            return correctionHint;
```
I don't think `correctionHint` belongs here. ActivityDefStore should only deal with caching the default validation message that corresponds to each type of validation rule.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/block/question/ValidationRuleCreator.java:
```diff
@@ -0,0 +1,166 @@
+        );
+    }
+
+    private String resolveRuleMessage(AIBuilderContext ctx, RuleDef ruleDef) {
```
I don't think we need to do any "resolving" here. If you look in the abstract parent `Rule` class, it will decide whether to use the `correctionHint` or the `defaultMessage`. So we just need to grab the default one from ActivityDefStore and pass it in.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/service/actvityinstancebuilder/block/question/PicklistCreatorHelper.java:
```diff
@@ -0,0 +1,61 @@
+        return picklistOption;
+    }
+
+    private void setGroupStableIdIfOptionBelongsToGroup(List<PicklistGroupDef> picklistGroupDefs, PicklistOption picklistOption) {
```
Might be better to have some sort of `Map<option_stable_id, group_def>` here instead of a `List` so we can turn this into `O(n)` (lookup in map is O(1) and we're doing N options) instead of `O(n^2)` (each lookup is O(n) and we're doing N options).

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/db/dto/ActivityInstanceDto.java:
```diff
@@ -28,6 +28,7 @@
     private Long firstCompletedAt;
     private boolean allowUnauthenticated;
     private int sectionIndex;
+    private Long editTimeoutSec;
```
I think we can remove this now?

yufengwng commented on pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/JdbiActivityInstance.sql.stg:
```diff
@@ -14,6 +14,7 @@ select ai.activity_instance_id,
        act.allow_unauthenticated,
        act.study_activity_code,
        act_type.activity_type_code as activity_type,
+       act.edit_timeout_sec,
```
Don't need this anymore since we're grabbing this from the activity definition.

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/definition/ComponentBlockDef.java:
```diff
@@ -26,6 +26,8 @@
     @SerializedName("hideNumber")
     private boolean hideNumber;
 
+    private transient long revisionId;
```
I got a bit confused on why we need this, and then I saw that this is being used in `FormBlockCreatorHelper`. Two things:
* Maybe we should rename this to `componentRevisionId`. I got confused on thinking that this is the "revision" of the block itself...
* If we inspect the code, we see that component revision id is not really used anywhere. Maybe there's a way to not require on this.

yufengwng reviewed (dismissed):
Have a few small comments. This looks good already, so let's get this merged and fix the small issues in a new PR.

Commit:
DDP-5402 - remove feature branch configuratio from CircleCI config

Commit:
DDP-5402 - fix in response to PR comments: remove unused fields; remove unused parameters; rename fields

yufengwng commented on pepper-apis/src/main/java/org/broadinstitute/ddp/db/dto/ActivityInstanceDto.java:
```diff
@@ -27,7 +27,6 @@
     private String sessionId;
     private Long firstCompletedAt;
     private boolean allowUnauthenticated;
-    private int sectionIndex;
```
Wait... why did we remove this one? Can we put section index back in?

Commit:
DDP-5402 - put back field section_index

yufengwng committed:
Merge branch 'develop' into DDP-5402-instance-data-read-optimization
